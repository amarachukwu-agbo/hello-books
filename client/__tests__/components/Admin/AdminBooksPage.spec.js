import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from 'enzyme';
import sinon from 'sinon';
import mockData from '../../__mocks__/mockData';
import ConnectedAdminBooksPage, { AdminBooksPage }
  from '../../../components/Admin/AdminBooksPage';

const props = {
  getBooks: sinon.spy(() => Promise.resolve()),
  deleteBook: sinon.spy(() => Promise.resolve()),
  editBook: sinon.spy(() => Promise.resolve()),
  isFetching: false,
  pagination: null,
};

const { $ } = global;

describe('<AdminBooksPage />', () => {
  afterEach(() => {
    global.$ = $;
  });
  it('calls componentDidMount()', () => {
    sinon.spy(AdminBooksPage.prototype, 'componentDidMount');
    mount(<AdminBooksPage {...props}/>);
    expect(AdminBooksPage.prototype.componentDidMount.calledOnce)
      .toEqual(true);
    expect(props.getBooks.calledOnce)
      .toEqual(true);
    expect(props.getBooks.calledWith(1));
  });

  it('renders correctly', () => {
    const wrapper = shallow(<AdminBooksPage {...props}/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls setBookForEdit()', () => {
    const openModalMock = jest.fn();
    global.$ = () => ({
      modal: openModalMock,
    });
    const wrapper = shallow(<AdminBooksPage {...props}/>);
    sinon.spy(wrapper.instance(), 'setBookForEdit');
    wrapper.instance().setBookForEdit(1, 0);
    expect(wrapper.instance().setBookForEdit.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().setBookForEdit.calledWith(1, 0));
    expect(wrapper.state().book.bookId).toEqual(1);
    expect(wrapper.state().book.bookIndex).toEqual(0);
  });

  it('calls the deleteBook() method', () => {
    const wrapper = shallow(<AdminBooksPage {...props}/>);
    sinon.spy(wrapper.instance(), 'deleteBook');
    wrapper.instance().deleteBook(1);
    expect(wrapper.instance().deleteBook.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().deleteBook.calledWith(1));
  });

  it('renders the preloader when books are fetching', () => {
    props.isFetching = true;
    const wrapper = shallow(<AdminBooksPage {...props}/>);
    expect(wrapper.find('Preloader')).toHaveLength(1);
  });

  it('render errors when there is an error fetching books', () => {
    props.error = 'Network Error';
    props.isFetching = false;
    const wrapper = shallow(<AdminBooksPage {...props}/>);
    expect(wrapper.find('h4').text())
      .toEqual('Oops! Couldn\'t fetch available books. Network Error');
  });

  it('renders books when books are available', () => {
    const { books } = mockData;
    props.errror = null;
    props.books = books.booksDataResponse.books;
    props.pagination = books.booksDataResponse.pagination;

    const wrapper = shallow(<AdminBooksPage {...props} />);
    expect(wrapper.find('AdminBooks')).toHaveLength(1);
    expect(wrapper.find('Pagination')).toHaveLength(1);
  });
});

describe('<ConnectedAdminBooksPage />', () => {
  it('should render successfully', () => {
    const mockStore = configureMockStore([thunk]);
    const store = mockStore({});
    const connectedWrapper = shallow(<ConnectedAdminBooksPage
      store = {store}/>);
    expect(connectedWrapper.length).toBe(1);
  });
});
