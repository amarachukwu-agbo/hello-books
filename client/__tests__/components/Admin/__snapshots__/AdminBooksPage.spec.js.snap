// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminBooksPage /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminBooksPage
    deleteBook={[Function]}
    editBook={[Function]}
    getBooks={[Function]}
    isFetching={false}
    pagination={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        undefined,
        <h5
          className="book-header center"
        >
          Books Catalog
        </h5>,
        <div
          className="divider"
        />,
        undefined,
        undefined,
      ],
      "className": "admin",
    },
    "ref": null,
    "rendered": Array [
      undefined,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Books Catalog",
          "className": "book-header center",
        },
        "ref": null,
        "rendered": "Books Catalog",
        "type": "h5",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "divider",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      undefined,
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          undefined,
          <h5
            className="book-header center"
          >
            Books Catalog
          </h5>,
          <div
            className="divider"
          />,
          undefined,
          undefined,
        ],
        "className": "admin",
      },
      "ref": null,
      "rendered": Array [
        undefined,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Books Catalog",
            "className": "book-header center",
          },
          "ref": null,
          "rendered": "Books Catalog",
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "divider",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        undefined,
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
